### PR TITLE
Avoid crash if collection closes while reminder is building

### PR DESCRIPTION
**Pull request by mikehardy**
https://couchdb.ankidroid.org/acralyzer/_design/acralyzer/index.html#/report-details/95855f86-64b7-4b29-be03-776c63135b1c

```
java.lang.RuntimeException: Unable to start receiver com.ichi2.anki.services.ReminderService: java.lang.IllegalStateException: Cannot perform this operation because the connection pool has been closed.
at android.app.ActivityThread.handleReceiver(ActivityThread.java:3194)
at android.app.ActivityThread.-wrap17(Unknown Source:0)
at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1672)
at android.os.Handler.dispatchMessage(Handler.java:106)
at android.os.Looper.loop(Looper.java:164)
at android.app.ActivityThread.main(ActivityThread.java:6494)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)
Caused by: java.lang.IllegalStateException: Cannot perform this operation because the connection pool has been closed.
at io.requery.android.database.sqlite.SQLiteConnectionPool.throwIfClosedLocked(SQLiteConnectionPool.java:2)
at io.requery.android.database.sqlite.SQLiteConnectionPool.waitForConnection(SQLiteConnectionPool.java:2)
at io.requery.android.database.sqlite.SQLiteConnectionPool.acquireConnection(SQLiteConnectionPool.java:1)
at io.requery.android.database.sqlite.SQLiteSession.acquireConnection(SQLiteSession.java:2)
at io.requery.android.database.sqlite.SQLiteSession.executeForCursorWindow(SQLiteSession.java:3)
at io.requery.android.database.sqlite.SQLiteQuery.fillWindow(SQLiteQuery.java:5)
at io.requery.android.database.sqlite.SQLiteCursor.fillWindow(SQLiteCursor.java:4)
at io.requery.android.database.sqlite.SQLiteCursor.getCount(SQLiteCursor.java:2)
at io.requery.android.database.AbstractCursor.moveToPosition(AbstractCursor.java:1)
at io.requery.android.database.AbstractCursor.moveToNext(AbstractCursor.java:1)
at com.ichi2.libanki.DB.queryScalar(DB.java:3)
at com.ichi2.libanki.DB.queryScalar(DB.java:1)
at com.ichi2.libanki.Sched._lrnForDeck(Sched.java:4)
at com.ichi2.libanki.Sched.deckDueList(Sched.java:10)
at com.ichi2.libanki.Sched.deckDueTree(Sched.java:1)
at com.ichi2.anki.services.ReminderService.onReceive(ReminderService.java:8)
at android.app.ActivityThread.handleReceiver(ActivityThread.java:3187)
```

This shows up in the crash database for people with auto-sync on startup,
that closes the database when it is done, at the same time the reminder could
still be in process of being built.

Simple extract method refactor for the part of reminder build that touches
the collection, and we give it one more try on reminder notification

If that doesn't work, we at least don't crash now

## How Has This Been Tested?

Difficult to reproduce because it is timing-related, but the alteration still functions correctly on the happy path, and by visual inspection I'm not sure how it could make things worse, while it should stop the crash

## Checklist
_Please, go through these checks before submitting the PR._

- [ ] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [ ] You have a descriptive commit message with a short title (first line, max 50 chars).
- [ ] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [ ] You have commented your code, particularly in hard-to-understand areas
- [ ] You have performed a self-review of your own code

### CODE REVIEW

**Comment by timrae** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity decreasing per file
==============================
+ AnkiDroid/src/main/java/com/ichi2/anki/services/ReminderService.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5009756&bid=16524687)

File: AnkiDroid/src/main/java/com/ichi2/anki/services/ReminderService.java [hunk 1]
```diff
@@ -23,11 +23,13 @@
 import androidx.core.app.NotificationCompat;
 import androidx.core.app.NotificationManagerCompat;
 import androidx.core.content.ContextCompat;
+import timber.log.Timber;
 
 import com.ichi2.anki.CollectionHelper;
 import com.ichi2.anki.IntentHandler;
 import com.ichi2.anki.NotificationChannels;
 import com.ichi2.anki.R;
+import com.ichi2.libanki.Collection;
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/anki/services/ReminderService.java]: Codacy Issue found: [Avoid unused imports such as 'com.ichi2.libanki.Collection'](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5009756)

**Comment by timrae** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity decreasing per file
==============================
+ AnkiDroid/src/main/java/com/ichi2/anki/services/ReminderService.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5009756&bid=16524687)

**Comment by timrae** [comment]
You'd probably at least want to sleep before recurring...?

**Comment by mikehardy** [comment]
> You'd probably at least want to sleep before recurring...?

I thought about that, but I think this is a one-shot deal, if sync is happening while the reminder is constructed and sync wins the race (whereupon it closes the collection when done before the reminder is built), then that only happens once because sync is done now.

So it seems fair to grab the collection again (which should cause it to initialize...) immediately, and I believe it will work as mDb.close()/mDb = null are sequential and the mDb == null is the check for Collection re-initialization?

These are bets on thread scheduling though, for safety's sake you're probably right. I will personally sleep on it then put a quick sleep in

**Comment by mikehardy** [comment]
Related: I just saw a second crash related to the sync killing off the connection pool (this time in StudyOptionsFragment). Maybe sync shouldn't close the database when it's done? That would require a lot of investigation but the question is: do you remember why sync closes the database? Any institutional knowledge would be illuminating. I can pontificate but you likely know

**Comment by mikehardy** [comment]
I see why it is necessary now based on previous PRs/comments. Closing/reopening the collection is the only way to make sure a rollback happens if there were errors, and for other collection consumers relying on a collection with errors is invalid so we can't "only close if errors" as they would have to handle that https://github.com/ankidroid/Anki-Android/pull/549

I think best for now is to re-try as in here to avoid crashes, but a re-think on how we handle sync-on-startup might be in order

I will push one more re-try commit that handles the StudyOptionsFragment case I saw crash

```
java.lang.IllegalStateException: Cannot perform this operation because the connection pool has been closed.
at io.requery.android.database.sqlite.SQLiteConnectionPool.throwIfClosedLocked(SQLiteConnectionPool.java:2)
at io.requery.android.database.sqlite.SQLiteConnectionPool.waitForConnection(SQLiteConnectionPool.java:2)
at io.requery.android.database.sqlite.SQLiteConnectionPool.acquireConnection(SQLiteConnectionPool.java:1)
at io.requery.android.database.sqlite.SQLiteSession.acquireConnection(SQLiteSession.java:2)
at io.requery.android.database.sqlite.SQLiteSession.executeForCursorWindow(SQLiteSession.java:3)
at io.requery.android.database.sqlite.SQLiteQuery.fillWindow(SQLiteQuery.java:5)
at io.requery.android.database.sqlite.SQLiteCursor.fillWindow(SQLiteCursor.java:4)
at io.requery.android.database.sqlite.SQLiteCursor.getCount(SQLiteCursor.java:2)
at io.requery.android.database.AbstractCursor.moveToPosition(AbstractCursor.java:1)
at io.requery.android.database.AbstractCursor.moveToNext(AbstractCursor.java:1)
at com.ichi2.libanki.DB.queryScalar(DB.java:3)
at com.ichi2.libanki.DB.queryScalar(DB.java:1)
at com.ichi2.libanki.Sched.haveBuried(Sched.java:3)
at com.ichi2.libanki.Sched.haveBuried(Sched.java:1)
at com.ichi2.anki.StudyOptionsFragment.configureToolbar(StudyOptionsFragment.java:19)
at com.ichi2.anki.StudyOptionsFragment.onCreateView(StudyOptionsFragment.java:10)
```

**Comment by timrae** [comment]
Sounds about right, that stuff is very complex and fragile, I don't remember the details like that anymore, definitely best to look at the commit messages and discussions on github for accurate information.

**Comment by mikehardy** [comment]
Agreed on both counts - hacky but quick win. I'm going to merge it and keep looking for quick crash fixes / finishing-and-releasing-nearly-ready-stuff until there are no more tall poppies to cut down (if that ever happens) then this would go in the design re-think pile.

**Comment by expatc** [comment]
I believe this is the issue I am currently running into with both auto-sync on startup and notifications enabled. The app was super slow and eventually crashed when trying to sync manually. I had to delete the app data to successfully disable both settings and now it appears working with just auto-sync turned on. I was getting this with every launch due to having a big deck (30k cards, with about 400 included in the notification). I'm in the alpha group and can test if you need it.

AnkiDroid Version = 2.9.4

Android Version = 9

ACRA UUID = 41915830-12c2-4bb5-8646-672b5e80ed72

**Comment by mikehardy** [comment]
@expatc - if this was definitely happening to you with 2.9.4 that is surprising - I should be able to find a crash by that ACRA UUID but I'm traveling the next week or so, and won't be able to investigate further. If you do the work to hop on the alpha release track we'll be able to collaborate on any eventual fix though, it will definitely land on alpha first.

Without wanting to harm your daily study, if you could turn on auto-sync at startup and reminders once on alpha, to make sure it reproduces that could help me save time

**Pull request merged**